proofs/src/program/http.rs

```diff
@@ -25,104 +25,167 @@ use crate::{
-  pub fn validate(&self) -> Result<(), ProofError> {
-    // TODO: What are legal statuses?
-    const VALID_STATUSES: [&str; 2] = ["200", "201"];
-    if !VALID_STATUSES.contains(&self.status.as_str()) {
-      return Err(ProofError::InvalidManifest("Unsupported HTTP status".to_string()));
-    }
+impl TryFrom<&[u8]> for NotaryResponseBody {
+  type Error = ProofError;
 
-    // TODO: What HTTP versions are supported?
-    if self.version != "HTTP/1.1" {
-      return Err(ProofError::InvalidManifest(
-        "Invalid HTTP version: ".to_string() + &self.version,
-      ));
+  fn try_from(body_bytes: &[u8]) -> Result<Self, Self::Error> {
+    if body_bytes.is_empty() {
+      return Ok(Self { json: None });
     }
+    // Attempt to parse the body as JSON.
+    let json: serde_json::Value = serde_json::from_slice(body_bytes)
+      .map_err(|_| ProofError::InvalidManifest("Failed to parse body as valid JSON".to_string()))?;
+    Ok(Self { json: Some(json) })
+  }
+}
 
-    // TODO: What is the max supported message length?
-    // TODO: Not covered by serde's #default annotation. Is '""' a valid message?
-    if self.message.len() > 1024 || self.message.is_empty() {
-      return Err(ProofError::InvalidManifest(
-        "Invalid message length: ".to_string() + &self.message,
-      ));
+impl NotaryResponseBody {
+  /// Verifies if the notary's JSON response matches the client-provided `json_path`.
+  ///
+  /// The `json_path` defines a sequence of keys for traversing the JSON structure.
+  /// Each key specifies either:
+  /// - A string to look for in a JSON object.
+  /// - An index to access an element in a JSON array.
+  /// - A concrete value at the end of the path.
+  ///
+  /// The function traverses the JSON step-by-step, returning `true` if all keys match,
+  /// or `false` otherwise.
+  fn matches_path(&self, json_path: &[JsonKey]) -> bool {
+    if json_path.is_empty() {
+      debug!("Invalid json_path: Path is empty");
+      return false;
     }
+    let Some(json) = &self.json else {
+      debug!("No JSON data to match against");
+      return false;
+    };
+    self.traverse_json_path(json, json_path)
+  }
 
-    // We always expect at least one header, "Content-Type"
-    if self.headers.len() > MAX_HTTP_HEADERS || self.headers.is_empty() {
-      return Err(ProofError::InvalidManifest(
-        "Invalid headers length: ".to_string() + &self.headers.len().to_string(),
-      ));
+  /// Traverses a JSON value using a specified path of keys and verifies if it matches the given
+  /// sequence.
+  ///
+  /// # Arguments
+  /// * `initial_value` - The root JSON value to start traversal from.
+  /// * `json_path` - A sequence of `JsonKey` elements defining the traversal path.
+  ///
+  /// # Returns
+  /// `true` if the traversal successfully matches the sequence or reaches the expected value;
+  /// `false` otherwise.
+  fn traverse_json_path(&self, initial_value: &serde_json::Value, json_path: &[JsonKey]) -> bool {
```

Comment:
does this mean Manifest will have to include the final value that it needs to match?

Reply:
have we added a test for this? can't see one

Reply:
Here: https://github.com/pluto/web-prover/pull/500/files#diff-453f720b0d5a493b4e39a988c2029a2b6dd2b9b7b69d558131ae99c307a5f3edR942-R955